app/ui/browser-modern/reducers/ui.js

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2016 Mozilla
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+import Immutable from 'immutable';
+
+import UIState from '../model/ui';
+import * as ActionTypes from '../constants/action-types';
+import * as EffectTypes from '../constants/effect-types';
+
+export default function(state = new UIState(), action) {
+  switch (action.type) {
+    case ActionTypes.SET_STATUS_TEXT:
+      return setStatusText(state, action.statusText);
+
+    case ActionTypes.SET_PAGE_SEARCH_VISIBILITY:
+      return setPageSearchVisibility(state, action.visibility);
+
+    case ActionTypes.SET_OVERVIEW_VISIBILITY:
+      return setOverviewVisibility(state, action.visibility);
+
+    case ActionTypes.CREATE_PAGE:
+    case ActionTypes.SET_SELECTED_PAGE:
+    case ActionTypes.SET_SELECTED_PAGE_INDEX:
+    case ActionTypes.SET_SELECTED_PAGE_PREVIOUS:
+    case ActionTypes.SET_SELECTED_PAGE_NEXT:
+      return state.withMutations(mut => {
+        setPageSearchVisibility(mut, false);
+        setOverviewVisibility(mut, false);
+      });
```

Comment:
It seems funny to use functions designed for immutability with a mutable object. Couldn't you do:

    state = setPageSearchVisibility(state, false);
    state = setOverviewVisibility(state, false);
    return state;